### PR TITLE
[SPARK-34225][CORE] Don't encode further when a URI form string is passed to addFile or addJar

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR fixes an issue that `addFile` and `addJar` further encode even though a URI form string is passed.
For example, the following operation will throw exception even though the file exists.
```
sc.addFile("file:/foo/test%20file.txt")
```

Another case is `--files` and `--jars` option when we submit an application.
```
bin/spark-shell --files "/foo/test file.txt"
```
The path above is transformed to URI form [here](https://github.com/apache/spark/blob/ecf4811764f1ef91954c865a864e0bf6691f99a6/core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala#L400) and passed to `addFile` so the same issue happens.

### Why are the changes needed?

This is a bug.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New test.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -1975,9 +1976,9 @@ class SparkContext(config: SparkConf) extends Logging {
     } else {
       val (keys, scheme) = if (path.contains("\\") && Utils.isWindows) {
         // For local paths with backslashes on Windows, URI throws an exception
-        (addLocalJarFile(new File(path)), "local")
+        (addLocalJarFile(convertPathToFileForWin(path)), "local")
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:1979]: @sarutak, I think here we assume the path is always a local path on Windows (not an URI) so I guess it's fine to leave it (?)

**Comment by HyukjinKwon** [comment]
Yeah, I agree with this change.

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 1 above)
Comment (sarutak) [reply core/src/main/scala/org/apache/spark/SparkContext.scala:1979]: Ah, seems so. if the path contains "\\", it's Windows' local path.
I'll leave it.

**Review by srowen — COMMENTED** [review]
Hm, are we sure? I'd expect that a user does not pass an encoded string. If it contains a `%` for example then it's really a `%` that needs encoding (or else, the input is incorrect).

**Comment by sarutak** [comment]
> Hm, are we sure? I'd expect that a user does not pass an encoded string. If it contains a % for example then it's really a % that needs encoding (or else, the input is incorrect).

If a URI form string contains an encoded character, it should be really an encoded character, while a non-URI form string contains an encoded character, it should be leave it.
This solution can keep `%` as it is.

`--files`
```
$ bin/spark-shell --files /path/to/test%file.txt
scala> sc.listFiles
res0: Seq[String] = Vector(file:///tmp/test%25file.txt)
```

`sc.addFile`
```
scala> sc.addFile("/path/to/test%file.txt")
scala> sc.listFiles
res0: Seq[String] = Vector(file:///tmp/test%25file.txt)
```

**Comment by srowen** [comment]
How do you know if it's a "real" encoded character or just looks like one? If a path actually contains "%25" then this would fail now.

**Comment by sarutak** [comment]
If an URI form string like `file:/foo/bar%25.txt` or `http://foo/bar%25.txt`, I think we can consider `%25` as an encoded character. URI can't contain non-encoded `%`.
If a non-URI form path actually contains `%25`, I think it should be considered it's not encoded.

**Comment by srowen** [comment]
OK I agree if it's specified as a URI, rather than a path

**Comment by sarutak** [comment]
Added new assertions for non-URI paths.

**Comment by sarutak** [comment]
retest please.

**Comment by sarutak** [comment]
retest this please.

**Comment by HyukjinKwon** [comment]
My only concern here is if there's any unexpected behaviour changes between `new Path(path)` and `Utils.resolveURI` because Hadoop's `Path` has not clearly defined behaviours in terms of local paths and URIs IIRC.

Still I agree that we're good as long as we can take canonical URIs and files in Spark, so should be fine to go ahead if we're sure on this.

**Comment by sarutak** [comment]
@HyukjinKwon OK, to minimize the risk, let's take care only the case which a path is an absolute URI and which contains `%`.

**Comment by srowen** [comment]
@sarutak does that mean you want to make further changes?

**Comment by sarutak** [comment]
@srowen No. The latest change is what I made to minimize the compatibility risk.

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
@HyukjinKwon @srowen Do you have any comment on the latest change?

**Review by srowen — APPROVED** [review]
LGTM, merge when ready

**Comment by sarutak** [comment]
Merged to `master` and `branch-3.1`. Thank you @srowen and @HyukjinKwon .

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 2]
```diff
@@ -1584,7 +1584,11 @@ class SparkContext(config: SparkConf) extends Logging {
...
     ): Unit = {
     val uri = if (!isArchive) {
-      new Path(path).toUri
+      if (Utils.isAbsoluteURI(path) && path.contains("%")) {
+        new URI(path)
+      } else {
+        new Path(path).toUri
+      }
     } else {
       Utils.resolveURI(path)
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:1593]: @sarutak, I think we can just repalce `Path` here too (looks like you're fixing this problem across the codebase, right?)

Comment (sarutak) [reply core/src/main/scala/org/apache/spark/SparkContext.scala:1593]: Yeah, the behavior of `Path` seems to be a little bit confusable and there may be lots of places where `Path` does not behave  as we expect.
O.K, I'll looking into it.

Comment (sarutak) [reply core/src/main/scala/org/apache/spark/SparkContext.scala:1593]: @HyukjinKwon BTW, Just replacing `Path` is my first idea of this PR.
https://github.com/apache/spark/pull/31718/commits/3c40f988de81600a75190751bf9987dfc93d09e9#diff-1e596e9e9bd50e3075847a66f8fecee741289b2b0478f9099c3bf2b680ebaa35R1586

But I changed to reflect [your comment](https://github.com/apache/spark/pull/31718#issuecomment-792225217).
Anyway, I'll open a followup PR.

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/SparkContext.scala:1593]: Yeah, I know :). Sorry for a bit of forth and back. If we change it all in the code base, I think it makes more sense to land a correct fix rather then staying safer and conservative with a bandaid fix

**Pull request closed without merging**